llmfoundry/callbacks/curriculum_learning_callback.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Enable curriculum learning by resuming with a different dataset.
+
+This callback is currently experimental. The API may change in the future.
+"""
+
+import logging
+from typing import Any, Dict, Iterable, Union
+
+from composer.core import Callback, State
+from composer.loggers import Logger
+from streaming import StreamingDataset
+from torch.utils.data import DataLoader
+
+log = logging.getLogger(__name__)
+
+
+class CurriculumLearning(Callback):
+    """Starts an epoch with a different dataset when resuming from a checkpoint.
+
+    This callback is currently experimental. The API may change in the future.
+
+    Args:
+        dataset_index (int): The index of the dataset currently being used.
+        current_dataset_config (Dict): The configuration of the dataset currently
+            being used.
+        dataset (StreamingDataset): The dataset currently being used.
+    """
+
+    def __init__(self, dataset_index: int, dataloader: Union[DataLoader,
+                                                             Iterable],
+                 current_dataset_config: Dict):
+        self.dataset_index = dataset_index
+        self.saved_dataset_index = 0
+        self.all_dataset_configs = []
+        # The current dataset config is resolved and passed in train.py
+        self.current_dataset_config = current_dataset_config
+        self.new_dataset_setup = False
+        self.dataset_config_appended = False
+
+        # Must pass in dataset directly since it is not actually accessible at Event.INIT in
+        # Composer. We need to get the new dataset state to override checkpoint dataset state.
+        # Check if we are using a DataLoader and StreamingDataset
+        if not isinstance(dataloader, DataLoader):
+            raise ValueError(
+                f'CurriculumLearning callback can only be used with a train ',
+                f'dataloader of type DataLoader, but got {type(dataloader)}.')
+        dataset = dataloader.dataset
+        if not isinstance(dataset, StreamingDataset):
+            raise ValueError(
+                f'CurriculumLearning callback only supports StreamingDataset ',
+                f'because it requires loading and saving dataset state. ',
+                f'Instead, got a dataset of type {type(dataset)}')
+        # Save the current dataset state so we can restore it if needed.
+        self.current_dataset_state = dataset.state_dict(0, False)
```

Comment:
I don't really understand what's going on here -- can you elaborate?

Reply:
We need to store dataset state like `num_canonical_nodes`, `initial_physical_nodes`, and `shuffle_seed` at init time because it gets overwritten by checkpoint loading. So if I'm resuming with a different dataset, the only time I'll have access to this info is at init time, and then I load it in in the `after_load` function

Reply:
Why can't I fetch this on event init? Isn't it accessible the same way?

Reply:
Sadly you can't, I tried it. The `State` obj in composer is created [here](https://github.com/mosaicml/composer/blob/12261d62335bab7e341735bad368a08eb3212cb7/composer/trainer/trainer.py#L1043), but the dataloader doesn't get set until [here](https://github.com/mosaicml/composer/blob/12261d62335bab7e341735bad368a08eb3212cb7/composer/trainer/trainer.py#L1300), which is after Event.INIT [here](https://github.com/mosaicml/composer/blob/12261d62335bab7e341735bad368a08eb3212cb7/composer/trainer/trainer.py#L1199).

Reply:
a little annoying. I suppose it's fine...

Reply:
indeed